CI: ci-integration: ubuntu-24.04-arm64: TestNodeManagerAbortReleaseIPReassignment

## CI failure

Looking at the CI dashboard, there are some recent instances of this flake.

```
--- FAIL: TestNodeManagerAbortReleaseIPReassignment (2.02s)

    logger.go:256: time=2025-10-14T03:27:10.181Z level=INFO source=/home/runner/work/cilium/cilium/pkg/ipam/node_manager.go:390 msg="Discovered new CiliumNode custom resource" name=node4
    logger.go:256: time=2025-10-14T03:27:10.181Z level=WARN source=/home/runner/work/cilium/cilium/pkg/ipam/node.go:688 msg="Unable to compute pending pods, will not surge-allocate" name=node4 instanceID="" error="pod store uninitialized"
    logger.go:256: time=2025-10-14T03:27:10.181Z level=INFO source=/home/runner/work/cilium/cilium/pkg/ipam/node.go:718 msg="Resolving IP deficit of node" name=node4 instanceID="" selectedInterface="" selectedPoolID=global maxIPsToAllocate=3 availableForAllocation=2048 emptyInterfaceSlots=0 available=0 used=0 neededIPs=3 remainingInterfaces=0
    logger.go:256: time=2025-10-14T03:27:10.191Z level=WARN source=/home/runner/work/cilium/cilium/pkg/ipam/node.go:688 msg="Unable to compute pending pods, will not surge-allocate" name=node4 instanceID="" error="pod store uninitialized"
    logger.go:256: time=2025-10-14T03:27:10.191Z level=INFO source=/home/runner/work/cilium/cilium/pkg/ipam/node.go:718 msg="Resolving IP deficit of node" name=node4 instanceID="" selectedInterface="" selectedPoolID=global maxIPsToAllocate=1 availableForAllocation=2045 emptyInterfaceSlots=0 available=3 used=3 neededIPs=1 remainingInterfaces=0
    logger.go:256: time=2025-10-14T03:27:12.202Z level=WARN source=/home/runner/work/cilium/cilium/pkg/ipam/node.go:923 msg="Unable to unassign IPs from interface" name=node4 instanceID="" available=4 used=2 excess=1 excessIPs=[4] releasing=[4] selectedInterface="" selectedPoolID=global error="unable to release IP 4: IP 4 not found" selectedInterface="" releasingAddresses=1
    node_manager_test.go:691: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/ipam/node_manager_test.go:691
        	Error:      	Received unexpected error:
        	            	unable to release IP 4: IP 4 not found
        	Test:       	TestNodeManagerAbortReleaseIPReassignment
```

* [Example one](https://github.com/cilium/cilium/actions/runs/18484350546/job/52665040599)
* [Example two](https://github.com/cilium/cilium/actions/runs/18401124523/job/52430400452)

@cilium/sig-ipam

--- a/pkg/ipam/node.go
+++ b/pkg/ipam/node.go
@@ -793,9 +793,8 @@ func (n *Node) abortNoLongerExcessIPs(excessMap map[string]bool) {
 }
 
 // handleIPReleaseResponse handles IPs agent has already responded to
+// caller must hold mutex lock
 func (n *Node) handleIPReleaseResponse(markedIP string, ipsToRelease *[]string) bool {
-	n.mutex.Lock()
-	defer n.mutex.Unlock()
 	if n.resource.Status.IPAM.ReleaseIPs != nil {
 		if status, ok := n.resource.Status.IPAM.ReleaseIPs[markedIP]; ok {
 			switch status {
@@ -811,12 +810,6 @@ func (n *Node) handleIPReleaseResponse(markedIP string, ipsToRelease *[]string) 
 		}
 	}
 	return false
-}
-
-func (n *Node) deleteLocalReleaseStatus(ip string) {
-	n.mutex.Lock()
-	defer n.mutex.Unlock()
-	delete(n.ipv4Alloc.ipReleaseStatus, ip)
 }
 
 // handleIPRelease implements IP release handshake needed for releasing excess IPs on the node.
@@ -835,6 +828,8 @@ func (n *Node) deleteLocalReleaseStatus(ip string) {
 func (n *Node) handleIPRelease(ctx context.Context, a *maintenanceAction) (instanceMutated bool, err error) {
 	var ipsToMark []string
 	var ipsToRelease []string
+
+	n.mutex.Lock()
 
 	// Update timestamps for IPs from this iteration
 	releaseTS := time.Now()
@@ -859,7 +854,7 @@ func (n *Node) handleIPRelease(ctx context.Context, a *maintenanceAction) (insta
 			// can be freed up. If the selected interface changes or if this IP is not excess anymore, remove entry
 			// from local maps.
 			delete(n.ipv4Alloc.ipsMarkedForRelease, markedIP)
-			n.deleteLocalReleaseStatus(markedIP)
+			delete(n.ipv4Alloc.ipReleaseStatus, markedIP)
 			continue
 		}
 		// Check if the IP release waiting period elapsed
@@ -874,7 +869,6 @@ func (n *Node) handleIPRelease(ctx context.Context, a *maintenanceAction) (insta
 		ipsToMark = append(ipsToMark, markedIP)
 	}
 
-	n.mutex.Lock()
 	for _, ip := range ipsToMark {
 		n.logger.Load().Debug(
 			"Marking IP for release",

--- a/pkg/ipam/node.go
+++ b/pkg/ipam/node.go
@@ -18,7 +18,6 @@ import (
 	"k8s.io/client-go/tools/cache"
 
 	operatorK8s "github.com/cilium/cilium/operator/k8s"
-	operatorOption "github.com/cilium/cilium/operator/option"
 	"github.com/cilium/cilium/operator/watchers"
 	"github.com/cilium/cilium/pkg/defaults"
 	"github.com/cilium/cilium/pkg/ipam/metrics"
@@ -132,6 +131,9 @@ type Node struct {
 
 	// logLimiter rate limits potentially repeating warning logs
 	logLimiter logging.Limiter
+
+	// ExcessIPReleaseDelay controls how long operator would wait before an IP previously marked as excess is released.
+	excessIPReleaseDelay time.Duration
 }
 
 // ipAllocAttrs represents IP-specific allocation attributes.
@@ -301,6 +303,10 @@ func (n *Node) getStaticIPTags() ipamTypes.Tags {
 // A negative number is returned to indicate release of addresses.
 func (n *Node) GetNeededAddresses() int {
 	stats := n.Stats()
+
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
 	if stats.IPv4.NeededIPs > 0 {
 		return stats.IPv4.NeededIPs
 	}
@@ -858,7 +864,7 @@ func (n *Node) handleIPRelease(ctx context.Context, a *maintenanceAction) (insta
 			continue
 		}
 		// Check if the IP release waiting period elapsed
-		if ts.Add(time.Duration(operatorOption.Config.ExcessIPReleaseDelay) * time.Second).After(time.Now()) {
+		if ts.Add(n.excessIPReleaseDelay).After(time.Now()) {
 			continue
 		}
 		// Handling for IPs we've already heard back from agent.
@@ -890,10 +896,13 @@ func (n *Node) handleIPRelease(ctx context.Context, a *maintenanceAction) (insta
 
 	if len(ipsToRelease) > 0 {
 		a.release.IPsToRelease = ipsToRelease
+
+		nodeStats := n.Stats()
+
 		scopedLog := n.logger.Load().With(
-			logfields.Available, n.stats.IPv4.AvailableIPs,
-			logfields.Used, n.stats.IPv4.UsedIPs,
-			logfields.Excess, n.stats.IPv4.ExcessIPs,
+			logfields.Available, nodeStats.IPv4.AvailableIPs,
+			logfields.Used, nodeStats.IPv4.UsedIPs,
+			logfields.Excess, nodeStats.IPv4.ExcessIPs,
 			logfields.ExcessIPs, a.release.IPsToRelease,
 			logfields.Releasing, ipsToRelease,
 			logfields.SelectedInterface, a.release.InterfaceID,
@@ -984,12 +993,17 @@ func (n *Node) maintainIPPool(ctx context.Context) (instanceMutated bool, err er
 	}
 
 	if len(n.getStaticIPTags()) > 0 {
-		if n.stats.IPv4.AssignedStaticIP == "" {
+		nodeStats := n.Stats()
+
+		if nodeStats.IPv4.AssignedStaticIP == "" {
 			ip, err := n.ops.AllocateStaticIP(ctx, n.getStaticIPTags())
 			if err != nil {
 				return false, err
 			}
+
+			n.mutex.Lock()
 			n.stats.IPv4.AssignedStaticIP = ip
+			n.mutex.Unlock()
 		}
 	}
 

--- a/pkg/ipam/node_manager.go
+++ b/pkg/ipam/node_manager.go
@@ -13,6 +13,7 @@ import (
 
 	"golang.org/x/sync/semaphore"
 
+	operatorOption "github.com/cilium/cilium/operator/option"
 	"github.com/cilium/cilium/pkg/backoff"
 	"github.com/cilium/cilium/pkg/controller"
 	ipamStats "github.com/cilium/cilium/pkg/ipam/stats"
@@ -300,6 +301,7 @@ func (n *NodeManager) Upsert(resource *v2.CiliumNode) {
 				ipsMarkedForRelease: make(map[string]time.Time),
 				ipReleaseStatus:     make(map[string]string),
 			},
+			excessIPReleaseDelay: time.Duration(operatorOption.Config.ExcessIPReleaseDelay) * time.Second,
 		}
 		node.logger.Store(node.rootLogger.With(fieldName, resource.Name))
 

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -700,15 +700,19 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 	node.instanceSync.Trigger()
 
 	require.Eventually(t, func() bool {
+		node.mutex.Lock()
+		defer node.mutex.Unlock()
+
 		status, exists := node.resource.Status.IPAM.ReleaseIPs[releasedIP]
 		return exists && string(status) == ipamOption.IPAMReadyForRelease
 	}, 10*time.Second, time.Second)
 
-	// Now simulate the operator releasing the IP and marking it as released
 	node.mutex.Lock()
+
+	// Now simulate the operator releasing the IP and marking it as released
 	delete(node.resource.Spec.IPAM.Pool, releasedIP)
 	node.resource.Status.IPAM.ReleaseIPs[releasedIP] = ipamOption.IPAMReleased
-	node.mutex.Unlock()
+
 	// Also mark it as released in the internal ipReleaseStatus map
 	node.ipv4Alloc.ipReleaseStatus[releasedIP] = ipamOption.IPAMReleased
 
@@ -718,6 +722,9 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 		node.resource.Spec.IPAM.Pool = ipamTypes.AllocationMap{}
 	}
 	node.resource.Spec.IPAM.Pool[releasedIP] = ipamTypes.AllocationIP{Resource: "eni-test"}
+
+	node.mutex.Unlock()
+
 	node.ops.AllocateIPs(context.Background(), &AllocationAction{
 		IPv4: IPAllocationAction{
 			AvailableForAllocation: 1,
@@ -729,6 +736,10 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 
 	require.Eventually(t, func() bool {
 		node.PopulateIPReleaseStatus(node.resource)
+
+		node.mutex.Lock()
+		defer node.mutex.Unlock()
+
 		_, inReleaseStatus := node.ipv4Alloc.ipReleaseStatus[releasedIP]
 		_, inMarkedForRelease := node.ipv4Alloc.ipsMarkedForRelease[releasedIP]
 		_, inReleaseIPs := node.resource.Status.IPAM.ReleaseIPs[releasedIP]

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -13,6 +13,7 @@ import (
 	"time"
 
 	"github.com/cilium/hive/hivetest"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
@@ -730,10 +731,13 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 		return !inReleaseStatus && !inMarkedForRelease && !inReleaseIPs
 	}, 10*time.Second, time.Second)
 
-	node = mngr.Get("node4")
-	require.NotNil(t, node)
-	require.Equal(t, 4, node.Stats().IPv4.AvailableIPs)
-	require.Equal(t, 2, node.Stats().IPv4.UsedIPs)
+	// Wait for the InstanceSync job to finish and update the node stats
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		node = mngr.Get("node4")
+		assert.NotNil(c, node)
+		assert.Equal(c, 4, node.Stats().IPv4.AvailableIPs)
+		assert.Equal(c, 2, node.Stats().IPv4.UsedIPs)
+	}, 15*time.Second, time.Second, "AvailableIPs and UsedIPs count did not update")
 }
 
 type nodeState struct {

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -689,7 +689,12 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 
 	// Run maintenance process again to process the acknowledgment
 	err = node.MaintainIPPool(context.Background())
-	require.NoError(t, err)
+
+	// Handle the case where the MaintainIPPool trigger runs first
+	expectedErrorStr := fmt.Sprintf("unable to release IP %s: IP %s not found", releasedIP, releasedIP)
+	require.Condition(t, func() bool {
+		return err == nil || err.Error() == expectedErrorStr
+	})
 
 	// Resync one more time to process acknowledgements.
 	node.instanceSync.Trigger()

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -680,9 +680,13 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 
 	node.PopulateIPReleaseStatus(node.resource)
 
+	node.mutex.Lock()
+
 	// Verify it's marked for release in the CiliumNode resource
 	require.Contains(t, node.resource.Status.IPAM.ReleaseIPs, releasedIP)
 	require.Equal(t, ipamOption.IPAMMarkForRelease, string(node.resource.Status.IPAM.ReleaseIPs[releasedIP]))
+
+	node.mutex.Unlock()
 
 	// Fake acknowledge IP for release like agent would
 	testipam.FakeAcknowledgeReleaseIps(node.resource)

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -33,7 +33,10 @@ var (
 	k8sapi = &k8sMock{}
 )
 
-const testPoolID = ipamTypes.PoolID("global")
+const (
+	testPoolID               = ipamTypes.PoolID("global")
+	testExcessIPReleaseDelay = 2
+)
 
 type allocationImplementationMock struct {
 	// mutex protects all fields of this structure
@@ -454,7 +457,7 @@ func TestNodeManagerMinAllocateAndPreallocate(t *testing.T) {
 // - PreAllocate 4
 // - MaxAboveWatermark 4
 func TestNodeManagerReleaseAddress(t *testing.T) {
-	operatorOption.Config.ExcessIPReleaseDelay = 2
+	operatorOption.Config.ExcessIPReleaseDelay = testExcessIPReleaseDelay
 	am := newAllocationImplementationMock()
 	require.NotNil(t, am)
 	mngr, err := NewNodeManager(hivetest.Logger(t), am, k8sapi, metricsmock.NewMockMetrics(), 10, true, false)
@@ -525,7 +528,7 @@ func TestNodeManagerReleaseAddress(t *testing.T) {
 // being resolved
 func TestNodeManagerAbortRelease(t *testing.T) {
 	var wg sync.WaitGroup
-	operatorOption.Config.ExcessIPReleaseDelay = 2
+	operatorOption.Config.ExcessIPReleaseDelay = testExcessIPReleaseDelay
 	am := newAllocationImplementationMock()
 	require.NotNil(t, am)
 	mngr, err := NewNodeManager(hivetest.Logger(t), am, k8sapi, metricsmock.NewMockMetrics(), 10, true, false)
@@ -602,7 +605,8 @@ func TestNodeManagerAbortRelease(t *testing.T) {
 // 3. Before being removed from status.ipam.release-ips, the same IP is reassigned to the pool (ie. AWS ENI in the case of AWS ENI IPAM)
 // 4. The IP is no longer considered excess, so the release handshake is aborted and the IP is deleted from status.ipam.release-ips
 func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
-	operatorOption.Config.ExcessIPReleaseDelay = 2
+	operatorOption.Config.ExcessIPReleaseDelay = testExcessIPReleaseDelay
+
 	am := newAllocationImplementationMock()
 	require.NotNil(t, am)
 	mngr, err := NewNodeManager(hivetest.Logger(t), am, k8sapi, metricsmock.NewMockMetrics(), 10, true, false)
@@ -651,7 +655,7 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 		}
 
 		for _, ts := range node.ipv4Alloc.ipsMarkedForRelease {
-			if ts.Add(time.Duration(operatorOption.Config.ExcessIPReleaseDelay) * time.Second).Before(time.Now()) {
+			if ts.Add(time.Duration(testExcessIPReleaseDelay) * time.Second).Before(time.Now()) {
 				return true
 			}
 		}
@@ -686,10 +690,10 @@ func TestNodeManagerAbortReleaseIPReassignment(t *testing.T) {
 	require.Contains(t, node.resource.Status.IPAM.ReleaseIPs, releasedIP)
 	require.Equal(t, ipamOption.IPAMMarkForRelease, string(node.resource.Status.IPAM.ReleaseIPs[releasedIP]))
 
-	node.mutex.Unlock()
-
 	// Fake acknowledge IP for release like agent would
 	testipam.FakeAcknowledgeReleaseIps(node.resource)
+
+	node.mutex.Unlock()
 
 	// Run maintenance process again to process the acknowledgment
 	err = node.MaintainIPPool(context.Background())

